docs/reference/mapping/types/keyword.asciidoc

```diff
@@ -11,8 +11,10 @@ The keyword family includes the following field types:
 addresses, hostnames, status codes, zip codes, or tags. 
 * <<constant-keyword-field-type,`constant_keyword`>> for keyword fields that always contain
 the same value.
-* <<wildcard-field-type,`wildcard`>>, which optimizes log lines and similar keyword values
-for grep-like <<query-dsl-wildcard-query,wildcard queries>>.
+* <<wildcard-field-type,`wildcard`>>, which optimizes high-cardinality keyword
+fields, such as log messages, for grep-like
+<<query-dsl-wildcard-query,`wildcard`>> and <<query-dsl-regexp-query,`regexp`>>
```

Comment:
>optimizes high-cardinality keyword fields, such as log messages,

My comment below is long and the TL/DR is that different people have different notions of what `keyword` fields are used for so it's perhaps unwise to make comparisons with it. We need to target advice to these different camps.

Part of the problem is we badged wildcard field as a better keyword field. I think in practice it's a better `text` field for non-human-authored large strings like log messages.

* In ECS they actually map log messages as a text field (which can be tough to query but fast).
* Some [3rd party security systems](https://socprime.com/blog/elastic-for-security-analysts-part-1-searching-strings/) mapped log messages as keyword fields because they were easier to query (but slow and missing large values).

So the assumption that log messages == keyword fields is not one that is universally shared.

Keyword fields are only like wildcard fields in that they don't chop strings into words - it's conceptually a single value. That is more about how they are implemented rather than how they are used. Classic keyword fields I tend to think of as short, structured values like country code or customer IDs. They are used to sort and aggregate on on and nearly always queried using exact **full**-values.

The [3rd party security systems](https://socprime.com/blog/elastic-for-security-analysts-part-1-searching-strings/) (ab)use of keyword fields for log messages was to overcome a **partial**-match search problem - not for their strengths in structured aggs, sorting or fast full-value-match search capabilities.

Why are wildcard fields more like a `text` field?  It's typically for large and unstructured content where you only want to search for bits of it (no aggs or sorting required). However, it's a different form of partial search to classic text-matching because you look for character sequences, not word sequences (there's no pre-determined opinion on what constitutes a "word" in the string because of Analyzers). 

In my view our previous advice that `wildcard == keyword` field was misleading and consequently the ECS team replaced many of their classic short, structured keyword fields with wildcard when it was unsuitable and this change had to be backed out. The disk costs and search speeds were worse than the classic keyword field. Ironically, the one field in ECS that might have benefitted from the change was the log `message` field and that stayed mapped as a `text` field. Changing this to the wildcard field however would be a big UX change - people would have to learn to search for character sequence`*foo*` rather than just the plain word `foo`. The problem with `foo` searches in `text` log messages was always that the tokenizer might consider `org.elastic.foo.bar` as one word and not match. _Word-based searches only work well with human language, and not so well for computer-speak._

Targeted advice:
For people considering switching from `text` to `wildcard` (like ECS ) we should document the necessary  change in UX and how word-based text matching like `quick AND brown AND fox` needs to change to equivalent queries for character sequences like `[Qq]uick\wbrown\wfox` (it doesn't help that KQL doesn't support regex currently).

For people like the 3rd party security folks who already use regexes and wildcard queries they have essentially been deploying a hack putting log messages into keyword fields because of their frustrations with text Analyzers. They need a different message (with wildcard field your query UX stays the same, but search speeds can be improved if log messages are many and varied plus we can deal with longer messages).

Reply:
Thanks for the explanation @markharwood. I'll work on an update to address this.